subsys/net/ip/net_if.c

```diff
@@ -505,9 +512,32 @@ enum net_verdict net_if_send_data(struct net_if *iface, struct net_pkt *pkt)
 		net_if_queue_tx(iface, pkt);
 	}
 
+	k_mutex_unlock(&lock);
+
 	return verdict;
 }
 
+int net_if_set_link_addr(struct net_if *iface,
+			 uint8_t *addr, uint8_t len,
+			 enum net_link_type type)
+{
+	if (net_if_flag_is_set(iface, NET_IF_UP)) {
```

Comment:
What if the interface is brought up before the mutex is taken?

Reply:
> What if the interface is brought up before the mutex is taken?

The network interface flag is atomic_t, so when reading the value, does it need mutex?

Reply:
No, but there's a TOCTOU race: the code checks that it's not up, so goes on to take the lock and set its address.  That's unprotected in a race with another thread does the same thing and also goes on to bring the interface up.  By the time the lock is taken here and the address is changed, the interface might already be running.

Or so it appears.

Reply:
> By the time the lock is taken here and the address is changed, the interface might already be running.

Fair enough, the lock can be moved few lines up, no problem with that.